python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,601 @@
+        raise RuntimeError("Cannot change the type of AboveMax")
+
+
+class BelowMin(Literal, Singleton):
+    def value(self):
+        raise RuntimeError("BelowMin has no value")
+
+    def to(self, type_var):
+        raise RuntimeError("Cannot change the type of BelowMin")
+
+
+class BooleanLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(BooleanLiteral, self).__init__(f"BooleanLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(BooleanType)
+    def _(self, type_var):
+        return self
+
+
+class IntegerLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(IntegerLiteral, self).__init__(f"IntegerLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(IntegerType)
+    def _(self, type_var):
+        return self
+
+    @to.register(LongType)
+    def _(self, type_var):
+        return LongLiteral(self.value)
+
+    @to.register(FloatType)
+    def _(self, type_var):
+        return FloatLiteral(self.value)
+
+    @to.register(DoubleType)
+    def _(self, type_var):
+        return DoubleLiteral(self.value)
+
+    @to.register(DateType)
+    def _(self, type_var):
+        return DateLiteral(self.value)
+
+    @to.register(DecimalType)
+    def _(self, type_var):
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value))
+        else:
+            return DecimalLiteral(
+                Decimal(self.value).quantize(
+                    Decimal("." + "".join(["0" for _ in range(1, type_var.scale)]) + "1"), rounding=ROUND_HALF_UP
+                )
+            )
+
+
+class LongLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(LongLiteral, self).__init__(f"LongLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(LongType)
+    def _(self, type_var):
+        return self
+
+    @to.register(IntegerType)
+    def _(self, type_var):
+        if IntegerType.max < self.value:
+            return AboveMax()
+        elif IntegerType.min > self.value:
+            return BelowMin()
+        return IntegerLiteral(self.value)
+
+    @to.register(FloatType)
+    def _(self, type_var):
+        return FloatLiteral(self.value)
+
+    @to.register(DoubleType)
+    def _(self, type_var):
+        return DoubleLiteral(self.value)
+
+    @to.register(TimeType)
+    def _(self, type_var):
+        return TimeLiteral(self.value)
+
+    @to.register(TimestampType)
+    def _(self, type_var):
+        return TimestampLiteral(self.value)
+
+    @to.register(DecimalType)
+    def _(self, type_var):
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value))
+        else:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP))
```

Comment:
I'm not sure this works. My testing shows that the `quantize` method will increase precision when used this way:

```python
type_var.scale
=> 2

decimal.Decimal(-932).quantize(decimal.Decimal((0, (1,), -type_var.scale)))
=> Decimal('-932.00')

decimal.Decimal(-932).quantize(decimal.Decimal((0, (1,), -type_var.scale))) / 10**type_var.scale
=> Decimal('-9.32')
```

Reply:
Ah, nevermind! I was thinking this was coming from an unscaled value. This works correctly when you use `Decimal("-9.32111")` and set the scale to 2.